src/lib.rs

```diff
@@ -0,0 +1,489 @@
+                                base_ids.insert(root);
+                                base_ids
+                            }
+                            None => workspace_members,
+                        }
+                    }
+                };
+
+                packages
+                    .iter()
+                    .filter(|p| !excluded_packages.contains(&meta[p].name))
+                    .copied()
+                    .collect()
+            }
+            ScopeMode::AllowList(patterns) => {
+                meta.packages
+                    .iter()
+                    // Deviating from cargo by not supporting patterns
+                    // Deviating from cargo by only checking workspace members
+                    .filter(|p| workspace_members.contains(&p.id) && patterns.contains(&p.name))
+                    .map(|p| &p.id)
+                    .collect()
+            }
+        };
+
+        meta.packages
+            .iter()
+            .filter(|&p| base_ids.contains(&p.id))
+            .collect()
+    }
+}
+
+impl Check {
+    pub fn new(current: Rustdoc) -> Self {
+        Self {
+            scope: Scope::default(),
+            current,
+            baseline: Rustdoc::from_registry_latest_crate_version(),
+            log_level: Default::default(),
+        }
+    }
+
+    pub fn with_package_selection(&mut self, selection: PackageSelection) -> &mut Self {
+        self.scope.mode = ScopeMode::DenyList(selection);
+        self
+    }
+
+    pub fn with_packages(&mut self, packages: Vec<String>) -> &mut Self {
+        self.scope.mode = ScopeMode::AllowList(packages);
+        self
+    }
+
+    pub fn with_baseline(&mut self, baseline: Rustdoc) -> &mut Self {
+        self.baseline = baseline;
+        self
+    }
+
+    pub fn with_log_level(&mut self, log_level: log::Level) -> &mut Self {
+        self.log_level = Some(log_level);
+        self
+    }
+
+    /// Some `RustdocSource`s don't contain a path to the project root,
+    /// so they don't have a target directory. We try to deduce the target directory
+    /// on a "best effort" basis -- when the source contains a target dir,
+    /// we use it, otherwise when the other source contains one, we use it,
+    /// otherwise we just use a standard cache folder as specified by XDG.
+    /// We cannot use a temporary directory, because the rustdocs from registry
+    /// are being cached in the target directory.
+    fn get_target_dir(&self, source: &RustdocSource) -> anyhow::Result<PathBuf> {
+        Ok(
+            if let Some(path) = get_target_dir_from_project_root(source)? {
+                path
+            } else if let Some(path) = get_target_dir_from_project_root(&self.current.source)? {
+                path
+            } else if let Some(path) = get_target_dir_from_project_root(&self.baseline.source)? {
+                path
+            } else {
+                get_xdg_cache_dir()
```

Comment:
I suspect this won't work on Windows, unfortunately. Let's instead use [the `directories` crate](https://crates.io/crates/directories) which is the known-working cross-platform way to do this:
```rust
if let Some(proj_dirs) = ProjectDirs::from("cargo-semver-checks") {
    proj_dirs.config_dir();
    // probable dirs (still need to be created! the library won't make them for us)
    // Lin: /home/alice/.config/cargo-semver-checks
    // Win: C:\Users\Alice\AppData\Roaming\cargo-semver-checks\config
    // Mac: /Users/Alice/Library/Application Support/cargo-semver-checks
}
```